### ISSUE TITLE
Get all versions of a class in a Multi-Release JAR

### ISSUE THREAD

**Issue by pkesseli** [classgraph/classgraph#787]
When invoking ClassGraph as follows:

```java
final ClassGraph classGraph = new ClassGraph().disableNestedJarScanning().ignoreClassVisibility();
try (final ScanResult scanResult = classGraph.verbose().scan()) {
    final ResourceList classes = scanResult.getAllResources();
    return classes.parallelStream().filter(resource -> resource.getPath().contains("StackLocator.class"))
            .map(resource -> resource.getPathRelativeToClasspathElement()).collect(Collectors.toList());
}
```

I only get the class `META-INF/versions/9/org/apache/logging/log4j/util/StackLocator.class`, not the base version `/org/apache/logging/log4j/util/StackLocator.class`. I would like to instrument all versions of the class, so how do I configure ClassGraph to iterate over all versions of classes?

Attached a minimal Maven project to reproduce the issue as well as the verbose ClassGraph output:
[classgraph-demo.tar.gz](https://github.com/classgraph/classgraph/files/12005752/classgraph-demo.tar.gz)
[classgraph.log](https://github.com/classgraph/classgraph/files/12005754/classgraph.log)

**Comment by lukehutch** [comment]
Hi, this is interesting -- by design, only a single classfile with a given package path (specifically the one that would be found in the current runtime environment by the running JVM) is returned by ClassGraph, when scanning classes. However, you are supposed to be able to find all versions of a resource, as your code attempts to do.

I made the assumption when I wrote the code that resources should also follow the `META-INF/versions` rules in the current runtime environment, if there are multiple versions available. The thinking here was that you should only be able to see one version of a resource *per package path, per classpath element*.

Before JVM 9, there could only be one version of a resource per package path, because the package path was defined relative to the root of the classpath element. Classfile versioning changed this, so you can have multiple versions of a resource within a single classpath element -- one per JVM version number, and one unversioned (legacy classpath) version.

Finding only the version that is relevant to the current runtime environment (the current behavior) is going to be the right thing to do for 99% of users. (Actually even the Java module system in the running JVM will not show you more versions of a resource than the one that works with the current JVM version...) But what we'll need to do for your usecase is disable version shadowing entirely.

I'm really tight on time right now -- do you have time to look into how to do this?

You would add a config option to `ClassGraph`, then write that option into the `ScanSpec`, as with the other options. Then search the ClassGraph source for `META-INF/versions` and `MULTI_RELEASE_PATH_PREFIX`. There are only a few pieces of code involved in this.

However, there may be some unintended consequences, e.g. for modules, paths should never start with `META-INF/versions`, since the Java module API does exactly what ClassGraph does here (remove the version prefix part of the path, then only return the first resource or class for a given path within the current module). I don't know what that would break (possibly only users' assumptions that ClassGraph should only return what the Java API returns for a given classpath element, using reflection and/or the module API). But as long as the ClassGraph tests pass with these moddfications, I guess you should be free to break convention here.

By the way you can get rid of `.ignoreClassVisibility()` if you are only scanning resources, not classes.

**Comment by pkesseli** [comment]
Hi Luke, thanks for your quick response and confirming that I wasn't just messing up my config. I think I can come up with a PR for this, I'll need to read myself into the ClassGraph source code first though.

```
possibly only users' assumptions that ClassGraph should only return what the Java API returns for a given classpath element, using reflection and/or the module API
```

I think it should be fine as long as this is hidden behind a documented flag like `new ClassGraph.enableMultiReleaseVersions()` and turned off by default.

**Comment by lukehutch** [comment]
Yes, sounds good. Thanks for taking a look!

**Comment by lukehutch** [comment]
Fixed and released in version 4.8.162. Thank you for the feature request and the pull request!
